contrib/md5/md5.1
.\"	BSDI	$Id: md5.1,v 1.1.1.1 1994/01/13 22:42:32 polk Exp $
.\" @(#)md5.1	6.1 1/10/94 22:52:41
.\"
.\" This man page was written by:
.\"
.\"	 Landon Curt Noll  (chongo@toad.com)	chongo <was here> /\../\
.\"
.\" This man page has been placed in the public domain.  Please do not 
.\" copyright this man page.
.\"
.\" LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH  REGARD  TO
.\" THIS  SOFTWARE,  INCLUDING  ALL IMPLIED WARRANTIES OF MER-
.\" CHANTABILITY AND FITNESS.  IN NO EVENT SHALL  LANDON  CURT
.\" NOLL  BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL
.\" DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM  LOSS  OF
.\" USE,  DATA  OR  PROFITS, WHETHER IN AN ACTION OF CONTRACT,
.\" NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR  IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" See md5drvr.c for version and modification history.
.\"
.TH MD5 1
.SH NAME
md5 \- RSA Data Security, Inc. MD5 Message-Digest Algorithm
.SH SYNOPSIS
.BR md5 \ [
.BR \-cdhiqtx \ ][
.BR \-p prefix\ ][
.BR \-P prfile\ ][
.BR \-s str\ ]
file\ ...
.SH DESCRIPTION
The
.B md5
utility implements the 
RSA Data Security, Inc. MD5 Message-Digest Algorithm (\fIMD5\fP).
It produces 128-bit MD5 Digests of files, strings or data 
read on stdin.
If no
.B file
is given, and
no string is to be digested (\fB\-s\fP)
then stdin will be digested.
By default, a digests are printed as 40 hex characters without a leading 
.IR 0x .
.PP
The
algorithm takes as input an input message of arbitrary length and
produces as output a 128-bit ``fingerprint'' or ``message digest'' of the
input. 
It is conjectured that it is computationally infeasible to
produce two messages having the same message digest, or to produce
any message having a given prespecified target message digest. The
.I MD5 
algorithm is intended for digital signature applications, where a
large file must be ``compressed'' in a secure manner before being
encrypted with a private (secret) key under a public-key cryptosystem
such as RSA.
.PP
If a 
.B str 
(string) argument is given, then the digest for 
.BR str ,
followed by a space, followed by
.B str 
enclosed double quotes is written to stdout.
Any
.B file
arguments are ignored.
.PP
If one or more
.B file
if given, a separate digest if produced for each file.
By default, 
.B file 
digests are followed by a space and the filename.
.PP
If no 
.B str
or
.B file
arguments are given, then a digest of stdin is written to stdout.
.PP
.TP
.B \-c
Print C style hex digests with a leading \fI0x\fP.
.TP
.B \-d
Compute dual digests.
.sp
Data is divided into two streams and digested separately.
Each successive octet is assigned an index starting with 0.
The even stream consists of octets with even indices.
The odd stream consists of octets with odd indices.
The even stream digest, followed by a space, followed by
the odd stream digest is written to stdout.
.TP
.B \-h
Print a help and usage message.
.TP
.B \-i
Compute inode digests of files.
.sp
Prepend the filename and various inode information to the file data 
being digested.
The inode information prepended includes information
such as the device (\fIst_dev\fP), 
inode number (\fIst_ino\fP), 
mode (\fIst_mode\fP), 
link count (\fIst_nlink\fP), 
uid (\fIst_uid\fP), 
gid (\fIst_gid\fP), 
size (\fIst_size\fP),
modification time (\fIst_mtime\fP) and
change time (\fIst_ctime\fP).
The prepended data is padded with zeros to make it a multiple of 64
bytes long.
Both a
.BR stat (2)
and a
.BR lstat (2)
information are both used.
.sp
The \fB\-i\fP flag allows one to include various inode information in
the digest.  
This option is useful in detecting file tampering.
For example, the following will produce different digests:
.sp
.in +0.5i
.nf
md5 \-i /tmp/chongo
cp /tmp/chongo /tmp/was_here
\^... misc funny business ...
cp /tmp/was_here /tmp/chongo
md5 \-i /tmp/chongo
.fi
.in -0.5i
.sp
Note that a \fB\-i\fP digest is portable to other systems.
This is because inode information will likely change as the
contents of a file are copied from machine to machine.
.sp
To mark a \fB\-i\fP digest as a special value, \fI.0\fP is 
prepended onto the digest output.
.sp
One may only compute inode digests of files.
Use of
.B \-i
disables reading from stdin.
The
.B \-i
flag is not compatible with
.BR \-s .
.TP
.BI \-p prefix
.TP
.BI \-P prfile
Insert a prefix into the data to be digested.
.sp
By using \fB\-p\fP\fIprefix\fP, one may prepend any set of data with a
string.  
The digest produced is equivalent to digest that is
produced with the string pretended to the data.  
Thus the following two commands produce the same digest:
.sp
.in +0.5i
.nf
md5 \-p curds \-s whey
md5 \-s curdswhey
.fi
.in -0.5i
.sp
By use of the \fB\-P\fP\fIprfile\fP interface, one may prepend using
up to the 32k of a file.  
This interface allows one to prepend using binary data.  
The following produces the same digest:
.sp
.in +0.5i
.nf
md5 \-P /usr/bin/awk /bin/ls
dd if=/usr/bin/awk of=/tmp/foo bs=32k count=1
cat /tmp/foo /bin/ls > /tmp/ls
md5 /tmp/ls
.fi
.in -0.5i
.sp
Knowledge of the original prepend data is only kept in
the digest.  
Thus one may use the prepend string as a ``salt'' making it intractable 
for someone else to reproduce the digest of a file without knowledge 
of the prepend data.
.sp
For sightly faster performance, use prepend data that is
a multiple of 64 bytes long.  
The easy way to do this is to use \fB-P\fP on a file that is at least 32k
bytes in length.
.sp
This feature works in conjunction all modes of operation except
the \fB\-t\fP and \fB\-x\fP modes.
.TP
.B \-q
Output only digests.
The filename or string will not be written to stdout.
.TP
.BR \-s str
Digest \fIstr\fP as if it were a string.
.sp
The trailing NUL byte is not digested.
No files are digested.
The \fIstr\fP is written to stdout enclosed in double quotes.
.TP
.BR \-t
Time the user cpu seconds needed to digest several megabytes of data.
The number of megabytes,
followed by the digest produced,
followed by the number of user cpu seconds
followed by the number of characters per user second is written to stdout.
.sp
On some systems with certain virtual memory characteristics,
you may need to run the performance test several times, ignoring the
first result.
The default amount of data tested is 16 megabytes.
On some systems, this amount may differ.
One may change the number of megabytes processed by changing the value of
TEST_MEG in the file md5drvr.c and recompiling.
.TP
.BR \-v
Print the version.
.TP
.BR \-x
Perform an extended standard SHS test suite.
.sp
The test suite will first look in the current directory for
.IR file1 .
If it is not found, it will look in 
.IR /usr/contrib/lib/md5 .
It is assumes that 
.I file2 
is in the same location as 
.sp
The standard test suite been extended.  
The initial string now says:
.sp
.in +0.5i
.nf
md5 test suite results
.fi
.in -0.5i
.sp
instead of:
.sp
.in +0.5i
.nf
SHA test suite results
.fi
.in -0.5i
.sp
to reflect the utility name instead of the algorithm implemented.
This also helps distinguish this version from much older 
versions which did not have the 
.B -v 
flag.
Also the original test file
.I foo
which contained the string  "\fIabc\fP" (with no newline) was renamed 
.IR file1 .
.SH "SEE ALSO"
.BR shs (1),
.BR stat (2),
.BR lstat (2)
.SH FILES
.nf
\^./file1			default test file location
\^./file2			default test file location
/usr/contrib/lib/md5/file1	alternate test file location
/usr/contrib/lib/md5/file2	alternate test file location
.fi
.SH AUTHOR
.nf
Much of this code was written, re-written or modified by:

    Landon Curt Noll	(chongo@toad.com)  /\\../\\

Part of this code code is based on code by Peter C. Gutmann.
.fi
.SH NOTICE
LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. 
IN NO EVENT SHALL LANDON CURT NOLL BE LIABLE FOR ANY SPECIAL, INDIRECT OR
CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
.PP
Parts of this code are:
.PP
Copyright (C) 1990, RSA Data Security, Inc. All rights reserved.
.PP
License to copy and use this software is granted provided that
it is identified as the "RSA Data Security, Inc. MD5 Message-
Digest Algorithm" in all material mentioning or referencing this
software or this function.
.PP
License is also granted to make and use derivative works
provided that such works are identified as "derived from the RSA
Data Security, Inc. MD5 Message-Digest Algorithm" in all
material mentioning or referencing the derived work.
.PP
RSA Data Security, Inc. makes no representations concerning
either the merchantability of this software or the suitability
of this software for any particular purpose.  It is provided "as
is" without express or implied warranty of any kind.
.PP
These notices must be retained in any copies of any part of this
documentation and/or software.
.SH BUGS
The command:
.sp
.in +0.5i
.nf
md5 -s '/\\"O/\\'
.fi
.in -0.5i
.sp
writes to stdout, the following line:
.sp
.in +0.5i
.nf
ae8b1f8db9eaa95693164bd224291da0 "/\\"O/\\"
.fi
.in -0.5i
.sp
The string written in double quotes by
.B \-s
may not be a valid string according to C or shell syntax.


contrib/md5/md5.c
/*	BSDI	$Id: md5.c,v 1.1.1.1 1994/01/13 22:42:32 polk Exp $	*/
/* @(#)md5.c	6.2 1/11/94 00:46:15 */
/*
 * md5 - RSA Data Security, Inc. MD5 Message-Digest Algorithm
 *
 * LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH  REGARD  TO
 * THIS  SOFTWARE,  INCLUDING  ALL IMPLIED WARRANTIES OF MER-
 * CHANTABILITY AND FITNESS.  IN NO EVENT SHALL  LANDON  CURT
 * NOLL  BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM  LOSS  OF
 * USE,  DATA  OR  PROFITS, WHETHER IN AN ACTION OF CONTRACT,
 * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR  IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * See md5drvr.c for version and modification history.
 */

/*
 ***********************************************************************
 ** Copyright (C) 1990, RSA Data Security, Inc. All rights reserved.  **
 **								      **
 ** License to copy and use this software is granted provided that    **
 ** it is identified as the "RSA Data Security, Inc. MD5 Message-     **
 ** Digest Algorithm" in all material mentioning or referencing this  **
 ** software or this function.					      **
 **								      **
 ** License is also granted to make and use derivative works	      **
 ** provided that such works are identified as "derived from the RSA  **
 ** Data Security, Inc. MD5 Message-Digest Algorithm" in all          **
 ** material mentioning or referencing the derived work.	      **
 **								      **
 ** RSA Data Security, Inc. makes no representations concerning       **
 ** either the merchantability of this software or the suitability    **
 ** of this software for any particular purpose.  It is provided "as  **
 ** is" without express or implied warranty of any kind.              **
 **								      **
 ** These notices must be retained in any copies of any part of this  **
 ** documentation and/or software.				      **
 ***********************************************************************
 */

/*
 ***********************************************************************
 **  Message-digest routines:					      **
 **  To form the message digest for a message M 		      **
 **    (1) Initialize a context buffer mdContext using MD5Init	      **
 **    (2) Call MD5Update on mdContext and M			      **
 **    (3) Call MD5Final on mdContext				      **
 **  The message digest is now in mdContext->digest[0...15]	      **
 ***********************************************************************
 */


#include "md5.h"
#ifdef __bsdi__
#include <machine/endian.h>
#else
#include "endian.h"
#include "align.h"
#endif

char *MD5_what="@(#)";	/* #(@) if checked in */

/* almost zero fill */
static BYTE PADDING[64] = {
    0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
};

/* 
 * The F, G, H and I are basic MD5 functions.  The following
 * identity saves one boolean operation.
 *
 * F: (((x) & (y)) | (~(x) & (z))) == ((z) ^ ((x) & ((y) ^ (z))))
 * G: (((x) & (z)) | ((y) & ~(z))) == ((y) ^ ((z) & ((x) ^ (y))))
 */
/* F, G, H and I are basic MD5 functions */
#define F(x, y, z) ((z) ^ ((x) & ((y) ^ (z))))
#define G(x, y, z) ((y) ^ ((z) & ((x) ^ (y))))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* rotate a 32 bit value */
#define ROT(X,n) (((X)<<(n)) | ((X)>>(32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4 */
/* Rotation is separate from addition to prevent recomputation */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define FF(a, b, c, d, x, s, ac) \
    {(a) += F((b), (c), (d)) + (x) + (ULONG)(ac); \
     (a) = ROT((a), (s)); \
     (a) += (b); \
    }
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define GG(a, b, c, d, x, s, ac) \
    {(a) += G((b), (c), (d)) + (x) + (ULONG)(ac); \
     (a) = ROT((a), (s)); \
     (a) += (b); \
    }
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define HH(a, b, c, d, x, s, ac) \
    {(a) += H((b), (c), (d)) + (x) + (ULONG)(ac); \
     (a) = ROT((a), (s)); \
     (a) += (b); \
    }
#define S41 6
#define S42 10
#define S43 15
#define S44 21
#define II(a, b, c, d, x, s, ac) \
    {(a) += I((b), (c), (d)) + (x) + (ULONG)(ac); \
     (a) = ROT((a), (s)); \
     (a) += (b); \
    }

/* forward declaration */
static void MD5Transform P((ULONG*, ULONG*));


/* 
 * MD5Init - initiale the message-digest context
 *
 * The routine MD5Init initializes the message-digest context
 * mdContext. All fields are set to zero.
 */
void 
MD5Init(mdContext)
    MD5_CTX *mdContext;
{
    /* load magic initialization constants */
    mdContext->buf[0] = (ULONG)0x67452301;
    mdContext->buf[1] = (ULONG)0xefcdab89;
    mdContext->buf[2] = (ULONG)0x98badcfe;
    mdContext->buf[3] = (ULONG)0x10325476;

    /* Initialise bit count */
    mdContext->countLo = mdContext->countHi = mdContext->datalen = 0L;
}


/* 
 * MD5Update - update message-digest context
 *
 * The routine MD5Update updates the message-digest context to
 * account for the presence of each of the characters inBuf[0..inLen-1]
 * in the message whose digest is being computed.
 */
void 
MD5Update(mdContext, inBuf, count)
    MD5_CTX *mdContext;
    BYTE *inBuf;
    UINT count;
{
    ULONG in[MD5_BLOCKSIZE/sizeof(ULONG)];
    ULONG datalen = mdContext->datalen;
    ULONG value;
    ULONG tmp;
    int cnt;

    /*
     * Update bitcount
     */
    tmp = mdContext->countLo;
    if ((mdContext->countLo = tmp + ((ULONG)count<<3)) < tmp ) {
        mdContext->countHi++;                             
    }
    mdContext->countHi += count >> 29;

    /*
     * Catch the case of a non-empty data inBuf
     */
    if (datalen > 0) {

	/* determine the size we need to copy */
	ULONG cpylen = MD5_BLOCKSIZE - datalen;

	/* case: new data will not fill the inBuf */
	if (cpylen > count) {
	    memcpy((char *)(mdContext->in_byte)+datalen, inBuf, count);
	    mdContext->datalen = datalen+count;
	    return;

	/* case: inBuf will be filled */
	} else {
	    memcpy((char *)(mdContext->in_byte)+datalen, inBuf, cpylen);
#if BYTE_ORDER == BIG_ENDIAN
	    /* byte swap data into little endian order */
	    for (cnt=0; cnt < MD5_BLOCKSIZE/sizeof(ULONG); ++cnt) {
		value = (mdContext->in_ulong[cnt] << 16) |
			(mdContext->in_ulong[cnt] >> 16);
		in[cnt] = ((value & 0xFF00FF00L) >> 8) |
			  ((value & 0x00FF00FFL) << 8);
	    }
	    MD5Transform(mdContext->buf, in);
#else
	    MD5Transform(mdContext->buf, mdContext->in_ulong);
#endif
	    inBuf += cpylen;
	    count -= cpylen;
	    mdContext->datalen = 0;
	}
    }

    /*
     * Process data in MD5_BLOCKSIZE chunks
     */
    while (count >= MD5_BLOCKSIZE) {
#if defined(MUST_ALIGN)
	memcpy(in, inBuf, MD5_BLOCKSIZE);
#if BYTE_ORDER == BIG_ENDIAN
	/* byte swap data into little endian order */
	for (cnt=0; cnt < MD5_BLOCKSIZE/sizeof(ULONG); ++cnt) {
	    value = (in[cnt] << 16) | (in[cnt] >> 16);
	    in[cnt] = ((value & 0xFF00FF00L) >> 8) |
		       ((value & 0x00FF00FFL) << 8);
	}
#endif
	MD5Transform(mdContext->buf, in);
#else
#if BYTE_ORDER == LITTLE_ENDIAN
	MD5Transform(mdContext->buf, (ULONG *)inBuf);
#else
	/* byte swap data into big endian order */
	for (cnt=0; cnt < MD5_BLOCKSIZE/sizeof(ULONG); ++cnt) {
	    value = (((ULONG *)inBuf)[cnt] << 16) | 
		    (((ULONG *)inBuf)[cnt] >> 16);
	    in[cnt] = ((value & 0xFF00FF00L) >> 8) |
		       ((value & 0x00FF00FFL) << 8);
	}
	MD5Transform(mdContext->buf, in);
#endif
#endif
	inBuf += MD5_BLOCKSIZE;
	count -= MD5_BLOCKSIZE;
    }

    /*
     * Handle any remaining bytes of data.
     * This should only happen once on the final lot of data
     */
    if (count > 0) {
	memcpy(mdContext->in_byte, inBuf, count);
    }
    mdContext->datalen = count;
}


/* 
 * MD5fullUpdate - update message-digest context with a full set of blocks
 *
 * The routine MD5Update updates the message-digest context to
 * account for the presence of each of the characters inBuf[0..inLen-1]
 * in the message whose digest is being computed.
 *
 * Unlike MD5Update, this routine will assume that the input length is
 * a mulitple of MD5_BLOCKSIZE bytes long.
 */
void 
MD5fullUpdate(mdContext, inBuf, count)
    MD5_CTX *mdContext;
    BYTE *inBuf;
    UINT count;
{
#if defined(MUST_ALIGN) || BYTE_ORDER == BIG_ENDIAN
    ULONG in[MD5_BLOCKSIZE/sizeof(ULONG)];
#endif
    ULONG datalen = mdContext->datalen;
    ULONG value;
    ULONG tmp;
    int cnt;

    /*
     * Update bitcount
     */
    tmp = mdContext->countLo;
    if ((mdContext->countLo = tmp + ((ULONG)count<<3)) < tmp ) {
        mdContext->countHi++;                             
    }
    mdContext->countHi += count >> 29;

    /*
     * Process data in MD5_BLOCKSIZE chunks
     */
    while (count >= MD5_BLOCKSIZE) {
#if defined(MUST_ALIGN)
	memcpy(in, inBuf, MD5_BLOCKSIZE);
#if BYTE_ORDER == BIG_ENDIAN
	/* byte swap data into little endian order */
	for (cnt=0; cnt < MD5_BLOCKSIZE/sizeof(ULONG); ++cnt) {
	    value = (in[cnt] << 16) | (in[cnt] >> 16);
	    in[cnt] = ((value & 0xFF00FF00L) >> 8) |
		       ((value & 0x00FF00FFL) << 8);
	}
#endif
	MD5Transform(mdContext->buf, in);
#else
#if BYTE_ORDER == LITTLE_ENDIAN
	MD5Transform(mdContext->buf, (ULONG *)inBuf);
#else
	/* byte swap data into little endian order */
	for (cnt=0; cnt < MD5_BLOCKSIZE/sizeof(ULONG); ++cnt) {
	    value = (((ULONG *)inBuf)[cnt] << 16) | 
		    (((ULONG *)inBuf)[cnt] >> 16);
	    in[cnt] = ((value & 0xFF00FF00L) >> 8) |
		       ((value & 0x00FF00FFL) << 8);
	}
	MD5Transform(mdContext->buf, in);
#endif
#endif
	inBuf += MD5_BLOCKSIZE;
	count -= MD5_BLOCKSIZE;
    }
}


/* 
 * MD5Final - terminate the message-digest computation
 *
 * The routine MD5Final terminates the message-digest computation and
 * ends with the desired message digest in mdContext->digest[0...15].
 */
void 
MD5Final(mdContext)
    MD5_CTX *mdContext;
{
    ULONG countLo, countHi;
    ULONG value;
    ULONG mdi = mdContext->datalen;
    UINT i, ii;
    UINT padLen;

    /* save number of bits */
    countLo = mdContext->countLo;
    countHi = mdContext->countHi;

    /* pad out to 56 mod 64 */
    padLen = (mdi < 56) ? (56 - mdi) : (120 - mdi);
    MD5Update(mdContext, PADDING, padLen);

    /* append length in bits and transform */
#if BYTE_ORDER == BIG_ENDIAN
    for (i = 0; i < 14; i++) {
	value = (mdContext->in_ulong[i] << 16) |
		(mdContext->in_ulong[i] >> 16);
	mdContext->in_ulong[i] = ((value & 0xFF00FF00L) >> 8) |
				    ((value & 0x00FF00FFL) << 8);
    }
#endif
    mdContext->in_ulong[14] = countLo;
    mdContext->in_ulong[15] = countHi;
    MD5Transform(mdContext->buf, mdContext->in_ulong);

    /* store buffer in digest */
#if BYTE_ORDER == LITTLE_ENDIAN
    memcpy(mdContext->digest, mdContext->buf, sizeof(mdContext->digest));
#else
    for (i = 0, ii = 0; i < 4; i++, ii += 4) {
	mdContext->digest[ii] = (BYTE)(mdContext->buf[i] & 0xFF);
	mdContext->digest[ii+1] = (BYTE)((mdContext->buf[i] >> 8) & 0xFF);
	mdContext->digest[ii+2] = (BYTE)((mdContext->buf[i] >> 16) & 0xFF);
	mdContext->digest[ii+3] = (BYTE)((mdContext->buf[i] >> 24) & 0xFF);
    }
#endif
}


/* 
 * Basic MD5 step. Transforms buf based on in.
 */
static void 
MD5Transform(buf, in)
    ULONG *buf;
    ULONG *in;
{
    ULONG a = buf[0], b = buf[1], c = buf[2], d = buf[3];

    /* Round 1 */
    FF( a, b, c, d, in[ 0], S11, 3614090360UL); /* 1 */
    FF( d, a, b, c, in[ 1], S12, 3905402710UL); /* 2 */
    FF( c, d, a, b, in[ 2], S13,  606105819UL); /* 3 */
    FF( b, c, d, a, in[ 3], S14, 3250441966UL); /* 4 */
    FF( a, b, c, d, in[ 4], S11, 4118548399UL); /* 5 */
    FF( d, a, b, c, in[ 5], S12, 1200080426UL); /* 6 */
    FF( c, d, a, b, in[ 6], S13, 2821735955UL); /* 7 */
    FF( b, c, d, a, in[ 7], S14, 4249261313UL); /* 8 */
    FF( a, b, c, d, in[ 8], S11, 1770035416UL); /* 9 */
    FF( d, a, b, c, in[ 9], S12, 2336552879UL); /* 10 */
    FF( c, d, a, b, in[10], S13, 4294925233UL); /* 11 */
    FF( b, c, d, a, in[11], S14, 2304563134UL); /* 12 */
    FF( a, b, c, d, in[12], S11, 1804603682UL); /* 13 */
    FF( d, a, b, c, in[13], S12, 4254626195UL); /* 14 */
    FF( c, d, a, b, in[14], S13, 2792965006UL); /* 15 */
    FF( b, c, d, a, in[15], S14, 1236535329UL); /* 16 */

    /* Round 2 */
    GG( a, b, c, d, in[ 1], S21, 4129170786UL); /* 17 */
    GG( d, a, b, c, in[ 6], S22, 3225465664UL); /* 18 */
    GG( c, d, a, b, in[11], S23,  643717713UL); /* 19 */
    GG( b, c, d, a, in[ 0], S24, 3921069994UL); /* 20 */
    GG( a, b, c, d, in[ 5], S21, 3593408605UL); /* 21 */
    GG( d, a, b, c, in[10], S22,   38016083UL); /* 22 */
    GG( c, d, a, b, in[15], S23, 3634488961UL); /* 23 */
    GG( b, c, d, a, in[ 4], S24, 3889429448UL); /* 24 */
    GG( a, b, c, d, in[ 9], S21,  568446438UL); /* 25 */
    GG( d, a, b, c, in[14], S22, 3275163606UL); /* 26 */
    GG( c, d, a, b, in[ 3], S23, 4107603335UL); /* 27 */
    GG( b, c, d, a, in[ 8], S24, 1163531501UL); /* 28 */
    GG( a, b, c, d, in[13], S21, 2850285829UL); /* 29 */
    GG( d, a, b, c, in[ 2], S22, 4243563512UL); /* 30 */
    GG( c, d, a, b, in[ 7], S23, 1735328473UL); /* 31 */
    GG( b, c, d, a, in[12], S24, 2368359562UL); /* 32 */

    /* Round 3 */
    HH( a, b, c, d, in[ 5], S31, 4294588738UL); /* 33 */
    HH( d, a, b, c, in[ 8], S32, 2272392833UL); /* 34 */
    HH( c, d, a, b, in[11], S33, 1839030562UL); /* 35 */
    HH( b, c, d, a, in[14], S34, 4259657740UL); /* 36 */
    HH( a, b, c, d, in[ 1], S31, 2763975236UL); /* 37 */
    HH( d, a, b, c, in[ 4], S32, 1272893353UL); /* 38 */
    HH( c, d, a, b, in[ 7], S33, 4139469664UL); /* 39 */
    HH( b, c, d, a, in[10], S34, 3200236656UL); /* 40 */
    HH( a, b, c, d, in[13], S31,  681279174UL); /* 41 */
    HH( d, a, b, c, in[ 0], S32, 3936430074UL); /* 42 */
    HH( c, d, a, b, in[ 3], S33, 3572445317UL); /* 43 */
    HH( b, c, d, a, in[ 6], S34,   76029189UL); /* 44 */
    HH( a, b, c, d, in[ 9], S31, 3654602809UL); /* 45 */
    HH( d, a, b, c, in[12], S32, 3873151461UL); /* 46 */
    HH( c, d, a, b, in[15], S33,  530742520UL); /* 47 */
    HH( b, c, d, a, in[ 2], S34, 3299628645UL); /* 48 */

    /* Round 4 */
    II( a, b, c, d, in[ 0], S41, 4096336452UL); /* 49 */
    II( d, a, b, c, in[ 7], S42, 1126891415UL); /* 50 */
    II( c, d, a, b, in[14], S43, 2878612391UL); /* 51 */
    II( b, c, d, a, in[ 5], S44, 4237533241UL); /* 52 */
    II( a, b, c, d, in[12], S41, 1700485571UL); /* 53 */
    II( d, a, b, c, in[ 3], S42, 2399980690UL); /* 54 */
    II( c, d, a, b, in[10], S43, 4293915773UL); /* 55 */
    II( b, c, d, a, in[ 1], S44, 2240044497UL); /* 56 */
    II( a, b, c, d, in[ 8], S41, 1873313359UL); /* 57 */
    II( d, a, b, c, in[15], S42, 4264355552UL); /* 58 */
    II( c, d, a, b, in[ 6], S43, 2734768916UL); /* 59 */
    II( b, c, d, a, in[13], S44, 1309151649UL); /* 60 */
    II( a, b, c, d, in[ 4], S41, 4149444226UL); /* 61 */
    II( d, a, b, c, in[11], S42, 3174756917UL); /* 62 */
    II( c, d, a, b, in[ 2], S43,  718787259UL); /* 63 */
    II( b, c, d, a, in[ 9], S44, 3951481745UL); /* 64 */

    buf[0] += a;
    buf[1] += b;
    buf[2] += c;
    buf[3] += d;
}
